### PR TITLE
WRLGS-11: Utils to timestamp stderr logs

### PR DESCRIPTION
**Pull request by BourgoisMickael**
Define functions that adds a timestamp to stderr logs: `uncaughtException` and `warning` events.

These functions should be used in every component, so if we want to refine the format of stderr logs, we can change only one function in one place
___

Based on assessment [S3C-8936_logs_inventory](https://docs.google.com/spreadsheets/d/1YnI1oCV9jwtmJpszkth4JcsdF9KOXSYACIomG3xWqdQ/edit?gid=0#gid=0)

The versions of werelogs used by S3C components between `S3C 7.10.8` and `S3C 9.4` are:

Current version | Next version | Comment
-- | -- | --
8.1.4 | 8.1.5 | Current default branch
8.1.0 | 8.1.0.1 | Need new branch `hotfix/8.1.0`
7.10.1.6 | 7.10.5 | `7.4.10` and `7.10.1.0` to `7.10.4.10`: they are all the same commit [`23dfe7c`](https://github.com/scality/werelogs/commit/23dfe7c60b2972c17d94ebf928c8dbf3b8837fbc)<br/>Branches `hotfix/7.4.10` = `hotfix/7.10.1` and above = `development/7.10`
7.10.3.9 | 7.10.5 | same as above
| | 8.0.1 | Existing branch `development/8.0` with unused version `8.0.0` in latest S3C assessed

___

> [!IMPORTANT]
> PR for `hotfix/8.1.0` :next_track_button:  https://github.com/scality/werelogs/pull/168 :previous_track_button: 

___

> [!NOTE]
> Logs will look like this for uncaught errors
>
> ```
> 2024-06-27T08:47:22.654Z: uncaughtException:
> Error: TestingError
>    at Object.<anonymous> (/home/mickael/scality/werelogs/tests/unit/fixtures/stderrUtils/catchUncaughtException.js:22:11)
>    at Module._compile (node:internal/modules/cjs/loader:1198:14)
>    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1252:10)
>    at Module.load (node:internal/modules/cjs/loader:1076:32)
>    at Function.Module._load (node:internal/modules/cjs/loader:911:12)
>    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
>    at node:internal/main/run_main_module:22:47
>```
> And like this for warnings
> ```
> (node:39843) DeprecationWarning: TestWarningMessage
>    at Object.<anonymous> (/home/mickael/scality/werelogs/tests/unit/fixtures/stderrUtils/catchWarning.js:15:17)
>    at Module._compile (node:internal/modules/cjs/loader:1198:14)
>    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1252:10)
>    at Module.load (node:internal/modules/cjs/loader:1076:32)
>    at Function.Module._load (node:internal/modules/cjs/loader:911:12)
>    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
>    at node:internal/main/run_main_module:22:47
> Above Warning Date: 2024-06-27T08:48:45.046Z
> ```
> ```
>(node:41812) NOTE: The AWS SDK for JavaScript (v2) will enter maintenance mode
>on September 8, 2024 and reach end-of-support on September 8, 2025.
>
>Please migrate your code to use AWS SDK for JavaScript (v3).
>For more information, check blog post at https://a.co/cUPnyil
>    at emitWarning (/home/mickael/scality/werelogs/node_modules/aws-sdk/lib/maintenance_mode_message.js:38:13)
>    at Timeout._onTimeout (/home/mickael/scality/werelogs/node_modules/aws-sdk/lib/maintenance_mode_message.js:46:5)
>    at listOnTimeout (node:internal/timers:559:17)
>    at processTimers (node:internal/timers:502:7)
>Above Warning Date: 2024-06-27T08:53:51.170Z
> ```
> I'll remove this warnings in components also

### CODE REVIEW

**Comment by BourgoisMickael** [comment]
ping

**Review by tmacro — APPROVED** [review]

**Review by dvasilas — APPROVED** [review]

File: lib/stderrUtils.js [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+function catchAndTimestampUncaughtException(
+    dateFct = defaultTimestamp, exitCode = 1,
+) {
+    process.on('uncaughtException', (err, origin) => {
+        printErrorWithTimestamp(err, origin, dateFct());
+        if (exitCode !== null) {
+            process.nextTick(() => process.exit(exitCode));
+        }
+    });
+}
```
Comment (ghost) [inline lib/stderrUtils.js:54]: Would it make sense not to stop the process if we pass a flag? Example: https://github.com/scality/vault2/blob/6a0c2cf3de174f5f339f1379227fbe87b0e99ce7/lib/vaultd.js#L30

Comment (BourgoisMickael) [reply lib/stderrUtils.js:54]: I think it should be the default behavior to stop the process.

Most processes don't have any listener on this.
I plan to use simply:
```
require('werelogs').stderrUtils.catchAndTimestampStderr();
```

Some processes have, like cloudserver when it does not exit all the time and could end in undefined behavior.

File: lib/stderrUtils.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+ * catchAndTimestampWarning();
+ * @param {Error} err see process event uncaughtException
+ * @param {uncaughtException|unhandledRejection} origin see process event
+ * @param {string} [date=`defaultTimestamp()`] Date to print
+ * @returns {boolean} see process.stderr.write
+ */
+function printErrorWithTimestamp(
+    err, origin, date = defaultTimestamp(),
+) {
+    return process.stderr.write(`${date}: ${origin}:\n${err.stack}\n`);
```
Comment (ghost) [inline lib/stderrUtils.js:29]: If we plan to crash the process, why not logging with `fatal` level here?

Comment (BourgoisMickael) [reply lib/stderrUtils.js:29]: Those are stderr logs, there is no `Logger` to format those logs as json.

I copied what's done there: https://github.com/scality/cloudserver/pull/5406/files#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346R7

And we decided for now we don't want to change the format completely, we want functions to encapsulate the logic and later if we need to we could change the format in those functions and just bump in all components

Comment (BourgoisMickael) [reply lib/stderrUtils.js:29]: We try not to change the behavior too much for now, we just need to include the timestamp in stderr for the CS

**Pull request merged**